garrysmod/lua/vgui/dhtml.lua

```diff
@@ -63,6 +63,73 @@ function PANEL:Call( js )
 	self:QueueJavascript( js )
 end
 
+local function BuildFunction( func, ... )
+
+	-- Build a Javascript-safe JS function call.
+	-- To be run by either panel:RunFunction or panel:QueueFunction
+	--
+	-- First parameter is the JS function name,
+	-- additional parameters are the values to provide to the function call.
+	-- The values provided to the function call are treated as strings except where noted below.
+	
+	local args, amount = table.Pack( ... )
+	
+	local formatArgs = {}
+	local safeArgs = {}
+	
+	for k, v in ipairs( args ) do
+	
+		if isbool( v ) then -- Boolean
+		
+			formatArgs[ k ] = '%s'
+			safeArgs[ k ] = v and true or false
+			
+		elseif isnumber( v ) then -- Numbers
+		
+			formatArgs[ k ] = '%s'
+			safeArgs[ k ] = v
+			
+		elseif IsColor( v ) then -- Colors, convert to CSS. Should be updated to use srlion's colour conversion PR when available.
+		
+			formatArgs[ k ] = '"%s"'
+			if v.a == 255 then -- don't include alpha if it isn't needed
+				safeArgs[ k ] = "rgb( " .. v.r .. ", " .. v.g .. ", " .. v.b .. " )"
+			else
+				safeArgs[ k ] = "rgba( " .. v.r .. ", " .. v.g .. ", " .. v.b .. ", " .. ( v.a / 255 ) .. " )"
+			end
+			safeArgs[ k ] = string.JavascriptSafe( safeArgs[ k ] )
+			
+		elseif istable( v ) then -- Tables, convert to json string
+		
+			formatArgs[ k ] = '"%s"'
+			safeArgs[ k ] = string.JavascriptSafe( util.TableToJSON( v ) )

```

Comment:
So far I can tell this is correctly implemented. It will result in a valid json string. When the function runs on the JS side it will arrive as an unescaped ready to use json string. `string.JavascriptSafe` makes it save to pass in `RunJavascript/QueueFunction`.